core/drivers/stm32_hsem.c

```diff
@@ -0,0 +1,283 @@
+				     HSEM_GpCIDCFGR_CONF_MASK);
+	}
+
+	/* Security and privilege RIF configuration */
+	io_clrsetbits32(hsem_d->base + HSEM_SECCFGR,
+			HSEM_SECCFGR_MASK & hsem_d->conf_data.access_mask[0],
+			hsem_d->conf_data.sec_conf[0]);
+	io_clrsetbits32(hsem_d->base + HSEM_PRIVCFGR,
+			HSEM_PRIVCFGR_MASK & hsem_d->conf_data.access_mask[0],
+			hsem_d->conf_data.priv_conf[0]);
+
+	if (!is_tdcid)
+		return;
+
+	/* Configure HSEM processors configuration registers */
+	for (i = 0; i < HSEM_NB_PROC; i++) {
+		/*
+		 * If a processor CID configuration is present, enable it.
+		 * Else, nothing to do.
+		 */
+		if (!hsem_d->rif_proc_conf[i])
+			continue;
+
+		io_clrsetbits32(hsem_d->base + HSEM_CnCIDCFGR(i + 1),
+				HSEM_CnCIDCFGR_CONF_MASK,
+				_CIDCFGR_CFEN | hsem_d->rif_proc_conf[i]);
+	}
+
+	/*
+	 * Configure HSEM groups configuration registers
+	 * If one semaphore is configured, all semaphores of its group
+	 * must be configured too and MUST have the same RIF
+	 * configuration.
+	 */
+	for (i = 0; i < HSEM_NB_SEM_GROUPS; i++) {
+		unsigned int hsem_idx = i * HSEM_NB_SEM_PER_GROUP;
+		unsigned int hsem_cid = 0;
+		unsigned int known_cid_idx = 0;
+
+		prev_cid_value = hsem_d->conf_data.cid_confs[hsem_idx];
+
+		/* If CID filtering is disabled, do nothing */
+		if (!(prev_cid_value & _CIDCFGR_CFEN))
+			continue;
+
+		/*
+		 * Check if configured CID corresponds to a processor's
+		 * CID in HSEM_CnCIDCFGR.
+		 */
+		for (j = 0; j < HSEM_NB_PROC; j++) {
+			uint32_t proc_cid = hsem_d->rif_proc_conf[j];
+
+			hsem_cid = (prev_cid_value & HSEM_CnCIDCFGR_SCID_MASK);
+			DMSG("hsem_cid %u", hsem_cid);
+			DMSG("proc_cid %u", proc_cid);
+			if (proc_cid == hsem_cid) {
+				known_cid_idx = BIT(j) <<
+						HSEM_GpCIDCFGR_SEM_WLIST_SHIFT;
+				break;
+			}
+		}
+		if (!known_cid_idx)
+			panic("Unknown HSEM processor CID");
+
+		/*
+		 * HSEM resources in the same group must have the same CID
+		 * filtering configuration. Else it is inconsistent.
+		 */
+		for (j = 0; j < HSEM_NB_SEM_PER_GROUP; j++)
+			if (hsem_d->conf_data.cid_confs[j + hsem_idx] !=
+			    prev_cid_value)
+				panic("Inconsistent HSEM RIF group config");
+
+		io_clrsetbits32(hsem_d->base + HSEM_GpCIDCFGR(i),
+				HSEM_GpCIDCFGR_CONF_MASK,
+				_CIDCFGR_CFEN | known_cid_idx);
+	}
+}
+
+static TEE_Result parse_dt(const void *fdt, int node)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t rif_conf = 0;
+	unsigned int i = 0;
+	int lenp = 0;
+	const fdt32_t *cuint = NULL;
+	struct dt_node_info info = { };
+	struct io_pa_va addr = { };
+
+	fdt_fill_device_info(fdt, &info, node);
+	assert(info.reg != DT_INFO_INVALID_REG &&
+	       info.reg_size != DT_INFO_INVALID_REG_SIZE);
+
+	addr.pa = info.reg;
+	hsem_d->base = io_pa_or_va_secure(&addr, info.reg_size);
+
+	/* Gate the IP */
+	res = clk_dt_get_by_index(fdt, node, 0, &hsem_d->hsem_clock);
+	if (res)
+		return res;
+
+	cuint = fdt_getprop(fdt, node, "st,protreg", &lenp);
+	if (!cuint)
+		panic("No RIF configuration available");
```

Comment:
Would it be better to print error message and return instead of panic here, as it is initial driver probe dt parsing.

Reply:
I can do that. IMO it's pretty much the same because in the end, the platform will panic. I'd rather panic right now to avoid further useless code execution.
Maybe @etienne-lms can give us a hint on what to do.

Reply:
Failing to apply a firewall an awaited configuration directive is, I think, I good reason to refuse to boot, hence panic.